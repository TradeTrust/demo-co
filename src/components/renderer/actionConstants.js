// Actions
export default {
  UPDATE_DOCUMENT: "UPDATE_DOCUMENT",
  DOCUMENT_TEMPLATE_REGISTER: "DOCUMENT_TEMPLATE_REGISTER",
  DOCUMENT_TEMPLATE_SELECT_TAB: "DOCUMENT_TEMPLATE_SELECT_TAB"
};
