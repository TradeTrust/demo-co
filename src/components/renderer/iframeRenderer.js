import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import connectToChild from 'penpal/lib/connectToChild';

import { getCertificate } from "../../reducers/certificate";
import { get } from "lodash";
import { certificateData } from "@tradetrust/tradetrust-certificate";

class IframeRenderer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      template: get(certificateData(props.document), "$template"),
      connection: null
    };
  }

  componentDidMount() {
    let iframe = document.querySelector("#frameless-iframe");
    let updateHeight = this.updateHeight.bind(this);
    let updateTemplates = this.updateTemplates.bind(this);
    this.state.connection = connectToChild({
      iframe,
      methods: {
        updateHeight,
        updateTemplates
      }
    });
    this.renderCertificate(this.props.document);
  }

  selectTemplateTab(i) {
    this.state.connection.promise.then(frame => frame.selectTemplateTab(i));
  }

  updateHeight(h) {
    // Adding 60 to account for extra height on firefox
    // this._iframe.height = h + 60;
  }

  updateTemplates(templates) {
    if (!templates) return;
    const templateSelector = document.getElementById("template-selectors");
    templateSelector.innerHTML = "";
    templates.forEach((t, i) => {
      const btn = document.createElement("button");
      btn.innerHTML = t.label;
      btn.id = `selector-${t.id}`;
      btn.onclick = () => this.selectTemplateTab(i);
      templateSelector.appendChild(btn);
    });
  }

  renderCertificate(cert) {
    this.state.connection.promise.then(frame => frame.renderCertificate(cert));
  }

  render() {
    const { template } = this.state;
    return (
      <>
        <div id="template-selectors" />

        <iframe
          title="Rendered Certificate"
          id="frameless-iframe"
          src={template.url}
          style={{ width: "100%", border: 0, height: "100%" }}
        />
      </>
    );
  }
}

const mapStateToProps = store => ({
  document: getCertificate(store)
});

export default connect(mapStateToProps)(IframeRenderer);

IframeRenderer.propTypes = {
  document: PropTypes.object
};
