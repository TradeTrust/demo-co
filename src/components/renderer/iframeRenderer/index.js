import React, { useEffect } from "react";
import PropTypes from "prop-types";
import connectToChild from "penpal/lib/connectToChild";
import { Store } from "store";

import {
  getDocument,
  getRenderType,
  registerTemplates as registerTemplatesAction
} from "components/renderer/actions";
import { certificateData } from "@govtechsg/tradetrust-schema";
import MultiDocRenderer from "../multiDocRenderer";

let connection = null;

const IframeRenderer = props => {
  const { state, dispatch } = React.useContext(Store);
  const renderType = getRenderType(state);
  const doc = getDocument(state);

  useEffect(() => {
    if (!doc) {
      return props.history.push("/");
    }

    let iframe = document.querySelector("iframe");
    connection = connectToChild({
      iframe,
      methods: {
        updateHeight,
        updateTemplates
      }
    });

    renderDocument(certificateData(doc));
  }, []);

  const selectTemplateTab = (conn => i => {
    conn.promise.then(frame => {
      frame.selectTemplateTab(i);
    });
  })(connection);

  const updateHeight = () => {
    // Adding 60 to account for extra height on firefox
    // this._iframe.height = h + 60;
  };

  const updateTemplates = templates => {
    if (!templates) return;
    registerTemplatesAction(dispatch, templates);
  };

  const renderDocument = doc => {
    connection.promise.then(frame => {
      frame.renderCertificate(doc);
    });
  };

  return (
    <>
      { doc && <MultiDocRenderer selectTemplateTab={i => selectTemplateTab(i)} />}
      <iframe
        title="Rendered Document"
        id="iframe"
        src={renderType ? renderType.url : ""}
        style={{ width: "100%", border: 0, height: "100%" }}
      />
    </>
  );
};

export default IframeRenderer;

IframeRenderer.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func
  })
};
