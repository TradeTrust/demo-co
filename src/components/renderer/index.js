import { get } from "lodash";
import { certificateData } from "@govtechsg/tradetrust-schema";
import constants from "./actionConstants";

export const initialState = {
  doc: null,
  docModified: null,
  renderType: null,
  templates: null,
  activeTemplateTab: 0
};

const {
  UPDATE_DOCUMENT,
  DOCUMENT_TEMPLATE_REGISTER,
  DOCUMENT_TEMPLATE_SELECT_TAB
} = constants;

// Action Creators
export function updateDocument(dispatch, payload) {
  return dispatch({
    type: UPDATE_DOCUMENT,
    payload
  });
}

export function getDocument(store) {
  return store.renderer.docModified;
}

export function getRenderType(store) {
  return store.renderer.renderType;
}

export function getTemplates(store) {
  return store.renderer.templates;
}

export function registerTemplates(dispatch, payload) {
  return dispatch({
    type: DOCUMENT_TEMPLATE_REGISTER,
    payload
  });
}

export function selectTemplateTab(dispatch, payload) {
  return dispatch({
    type: DOCUMENT_TEMPLATE_SELECT_TAB,
    payload
  });
}

export function getActiveTemplateTab(store) {
  return store.renderer.activeTemplateTab;
}

// Reducers
export default function reducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_DOCUMENT:
      return {
        ...initialState,
        doc: action.payload,
        docModified: action.payload,
        renderType: get(certificateData(action.payload), "$template")
      };
    case DOCUMENT_TEMPLATE_REGISTER:
      return {
        ...state,
        templates: action.payload,
        activeTemplateTab: 0
      };
    case DOCUMENT_TEMPLATE_SELECT_TAB:
      return {
        ...state,
        activeTemplateTab: action.payload
      };
    default:
      return state;
  }
}
