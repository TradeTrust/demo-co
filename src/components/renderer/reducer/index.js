import { types, initialState } from "./constants";

const {
  DOCUMENT_TEMPLATE_REGISTER,
  DOCUMENT_TEMPLATE_SELECT_TAB
} = types;

// Reducers
export default function reducer(state = initialState, action) {
  switch (action.type) {
    case DOCUMENT_TEMPLATE_REGISTER:
      return {
        ...state,
        templates: action.payload,
        activeTemplateTab: 0
      };
    case DOCUMENT_TEMPLATE_SELECT_TAB:
      return {
        ...state,
        activeTemplateTab: action.payload
      };
    default:
      return state;
  }
}
