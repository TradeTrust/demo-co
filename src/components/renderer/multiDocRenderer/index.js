import React, { useState } from "react";
import PropTypes from "prop-types";

import { getTemplates } from "components/renderer";
import { Store } from "store";

const MultiDocRenderer = ({ selectTemplateTab }) => {
  const [activeTab, setActiveTab] = useState(0);
  const { state } = React.useContext(Store);
  const templates = getTemplates(state);
  return (
    <ul className="nav nav-tabs container" id="myTab" role="tablist">
      {templates && templates.length > 0
        ? templates.map((t, idx) => (
            <li key={idx} className="nav-item">
              <a
                className={`nav-link ${idx === activeTab ? "active" : ""}`}
                id={t.id}
                onClick={() => {
                  selectTemplateTab(idx);
                  setActiveTab(idx);
                }}
                role="tab"
                aria-controls="home"
                aria-selected="true"
              >
                {t.label}
              </a>
            </li>
          ))
        : null}
    </ul>
  );
};

export default MultiDocRenderer;

MultiDocRenderer.propTypes = {
  document: PropTypes.object,
  templates: PropTypes.array,
  selectTemplateTab: PropTypes.func
};
