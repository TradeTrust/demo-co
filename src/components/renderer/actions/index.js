import { types } from "components/renderer/reducer/constants";

const { DOCUMENT_TEMPLATE_REGISTER, DOCUMENT_TEMPLATE_SELECT_TAB } = types;

export function getDocument(store) {
  return store.renderer.docModified;
}

export function getRenderType(store) {
  return store.renderer.renderType;
}

export function getTemplates(store) {
  return store.renderer.templates;
}

export function registerTemplates(dispatch, payload) {
  return dispatch({
    type: DOCUMENT_TEMPLATE_REGISTER,
    payload
  });
}

export function selectTemplateTab(dispatch, payload) {
  return dispatch({
    type: DOCUMENT_TEMPLATE_SELECT_TAB,
    payload
  });
}

export function getActiveTemplateTab(store) {
  return store.renderer.activeTemplateTab;
}
