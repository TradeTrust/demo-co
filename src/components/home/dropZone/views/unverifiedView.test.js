import React from "react";
import UnVerifiedView from "./unverifiedView";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { shallow } from "enzyme";

const hashStatus = { verified: true },
  issuedStatus = { verified: true },
  notRevokedStatus = { verified: true };

const hashStatusFalse = { verified: false },
  issuedStatusFalse = { verified: false },
  notRevokedStatusFalse = { verified: false };

it("unverified view renders correctly when hash status, issued status and not revoked status is true", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UnVerifiedView
          hashStatus={hashStatus}
          issuedStatus={issuedStatus}
          notRevokedStatus={notRevokedStatus}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("unverified view renders correctly when hash status, issued status and not revoked status is false", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UnVerifiedView
          hashStatus={hashStatusFalse}
          issuedStatus={issuedStatusFalse}
          notRevokedStatus={notRevokedStatusFalse}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("unverified view renders correctly when hash status is false, issue status is true and not revoked status is true", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UnVerifiedView
          hashStatus={hashStatusFalse}
          issuedStatus={issuedStatus}
          notRevokedStatus={notRevokedStatus}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("unverified view renders correctly when hash status is true, issue status is false and not revoked status is true", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UnVerifiedView
          hashStatus={hashStatus}
          issuedStatus={issuedStatusFalse}
          notRevokedStatus={notRevokedStatus}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
it("unverified view renders correctly when hash status is true, issue status is true and not revoked status is false", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <UnVerifiedView
          hashStatus={hashStatus}
          issuedStatus={issuedStatus}
          notRevokedStatus={notRevokedStatusFalse}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Unverified View component", () => {
  it("Test click event", () => {
    const resetData = jest.fn();
    const view = shallow(
      <UnVerifiedView
        hashStatus={hashStatus}
        issuedStatus={issuedStatus}
        notRevokedStatus={notRevokedStatusFalse}
        resetData={resetData}
      />
    );
    view.find("button").simulate("click", { preventDefault: () => {} });
    expect(resetData).toHaveBeenCalledTimes(1);
  });
});
