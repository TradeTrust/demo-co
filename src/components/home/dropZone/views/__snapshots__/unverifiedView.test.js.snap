// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unverified view renders correctly when hash status is false, issue status is true and not revoked status is true 1`] = `
<div
  className="viewer-container invalid"
  style={
    Object {
      "backgroundColor": "#fbeae9",
      "borderRadius": 10,
    }
  }
>
  <span
    className="message-container"
  >
    <img
      alt="document invalid"
      src="test-file-stub"
    />
    <span
      className="invalid m-3"
      style={
        Object {
          "fontSize": "1.5rem",
        }
      }
    >
      This document is not valid
    </span>
  </span>
  <div
    className="verifications"
  >
    <p
      className="messages"
    >
      The document's contents are inaccurate
    </p>
  </div>
  <a
    href="/faq"
    onClick={[Function]}
  >
    <div
      className="unverified-btn"
    >
      What should I do?
    </div>
  </a>
  <div
    className="secondary-links"
  >
    <span>
      <a
        href="/ "
        onClick={[Function]}
      >
        <button
          className="text-link"
          onClick={[Function]}
        >
          Try another
        </button>
      </a>
    </span>
    
  </div>
</div>
`;

exports[`unverified view renders correctly when hash status is true, issue status is false and not revoked status is true 1`] = `
<div
  className="viewer-container invalid"
  style={
    Object {
      "backgroundColor": "#fbeae9",
      "borderRadius": 10,
    }
  }
>
  <span
    className="message-container"
  >
    <img
      alt="document invalid"
      src="test-file-stub"
    />
    <span
      className="invalid m-3"
      style={
        Object {
          "fontSize": "1.5rem",
        }
      }
    >
      This document is not valid
    </span>
  </span>
  <div
    className="verifications"
  >
    <p
      className="messages"
    >
      The document records are not found
    </p>
  </div>
  <a
    href="/faq"
    onClick={[Function]}
  >
    <div
      className="unverified-btn"
    >
      What should I do?
    </div>
  </a>
  <div
    className="secondary-links"
  >
    <span>
      <a
        href="/ "
        onClick={[Function]}
      >
        <button
          className="text-link"
          onClick={[Function]}
        >
          Try another
        </button>
      </a>
    </span>
    
  </div>
</div>
`;

exports[`unverified view renders correctly when hash status is true, issue status is true and not revoked status is false 1`] = `
<div
  className="viewer-container invalid"
  style={
    Object {
      "backgroundColor": "#fbeae9",
      "borderRadius": 10,
    }
  }
>
  <span
    className="message-container"
  >
    <img
      alt="document invalid"
      src="test-file-stub"
    />
    <span
      className="invalid m-3"
      style={
        Object {
          "fontSize": "1.5rem",
        }
      }
    >
      This document is not valid
    </span>
  </span>
  <div
    className="verifications"
  >
    <p
      className="messages"
    >
      The document has been revoked
    </p>
  </div>
  <a
    href="/faq"
    onClick={[Function]}
  >
    <div
      className="unverified-btn"
    >
      What should I do?
    </div>
  </a>
  <div
    className="secondary-links"
  >
    <span>
      <a
        href="/ "
        onClick={[Function]}
      >
        <button
          className="text-link"
          onClick={[Function]}
        >
          Try another
        </button>
      </a>
    </span>
    
  </div>
</div>
`;

exports[`unverified view renders correctly when hash status, issued status and not revoked status is false 1`] = `
<div
  className="viewer-container invalid"
  style={
    Object {
      "backgroundColor": "#fbeae9",
      "borderRadius": 10,
    }
  }
>
  <span
    className="message-container"
  >
    <img
      alt="document invalid"
      src="test-file-stub"
    />
    <span
      className="invalid m-3"
      style={
        Object {
          "fontSize": "1.5rem",
        }
      }
    >
      This document is not valid
    </span>
  </span>
  <div
    className="verifications"
  >
    <p
      className="messages"
    >
      The document's contents are inaccurate
    </p>
    <p
      className="messages"
    >
      The document records are not found
    </p>
    <p
      className="messages"
    >
      The document has been revoked
    </p>
  </div>
  <a
    href="/faq"
    onClick={[Function]}
  >
    <div
      className="unverified-btn"
    >
      What should I do?
    </div>
  </a>
  <div
    className="secondary-links"
  >
    <span>
      <a
        href="/ "
        onClick={[Function]}
      >
        <button
          className="text-link"
          onClick={[Function]}
        >
          Try another
        </button>
      </a>
    </span>
    
  </div>
</div>
`;

exports[`unverified view renders correctly when hash status, issued status and not revoked status is true 1`] = `
<div
  className="viewer-container warning"
  style={
    Object {
      "backgroundColor": "#fbf6e9",
      "borderRadius": 10,
    }
  }
>
  <span
    className="message-container"
  >
    <img
      alt="document warning"
      src="test-file-stub"
    />
    <span
      className="warning m-3"
      style={
        Object {
          "fontSize": "1.5rem",
        }
      }
    >
      Document from unregistered institution
    </span>
  </span>
  <div
    className="verifications"
  />
  <a
    href="/faq"
    onClick={[Function]}
  >
    <div
      className="unverified-btn"
    >
      What should I do?
    </div>
  </a>
  <div
    className="secondary-links"
  >
    <span>
      <a
        href="/ "
        onClick={[Function]}
      >
        <button
          className="text-link"
          onClick={[Function]}
        >
          Try another
        </button>
      </a>
    </span>
    <span
      onClick={[Function]}
    >
      <a
        href="/viewer"
        onClick={[Function]}
      >
        <button
          className="text-link"
          id="certificate-view-anyway"
        >
          View document anyway
        </button>
      </a>
    </span>
  </div>
</div>
`;
